File: Medium/Possible Bipartition/possible-bipartition.cpp
class Solution {
public:
    bool possibleBipartition(int n, vector<vector<int>>& dislikes) {
        const int BLANK = 0;
        const int BLUE = 10e4;
        const int RED = -10e4;

        vector<int> m(n, BLANK);
        map<int, vector<int>> graph;

        for (auto d : dislikes)
        {
            int k = d[0] - 1;
            int v = d[1] - 1;
            graph[k].emplace_back(v);
            graph[v].emplace_back(k);
        }

        function<bool(int, int)> dfs = [&](int vertex, int color)
        {
            m[vertex] = color;

            for (auto v : graph[vertex])
            {
                if (m[v] == color) return false;
                if (m[v] == BLANK && !dfs(v, -color))
                    return false;
            }
            return true;
        };

        for (int i = 0; i < n; ++i)
        {
            if (m[i] == BLANK && !dfs(i, BLUE))
                return false;
        }

        return true;
    }
};

File: Medium/Possible Bipartition/possible-bipartition.py
class Solution:
    def possibleBipartition(self, n: int, dislikes: List[List[int]]) -> bool:

        m = [0] * n
        blank = 0
        blue = 10e3
        red = -10e3
        graph = defaultdict(list)

        for d in dislikes:
            graph[d[0] - 1].append(d[1] - 1)
            graph[d[1] - 1].append(d[0] - 1)

        def dfs(vertex, color):
            m[vertex] = color
            for v in graph[vertex]:
                if m[v] == color: return False
                if m[v] == blank and not dfs(v, -color):
                    return False
            return True

        
        for i in range(n):
            if m[i] == blank and not dfs(i, blue):
                return False

        return True
